src/main/java/org/broadinstitute/hellbender/utils/nio/NioFileCopier.java

```diff
@@ -0,0 +1,624 @@
+package org.broadinstitute.hellbender.utils.nio;
+
+import org.apache.commons.collections4.queue.CircularFifoQueue;
+import org.apache.commons.io.FileUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+
+/**
+ * Class to copy a file using {@link java.nio}.
+ * Operates using paths.
+ *
+ * INSTANCES OF THIS CLASS ARE NOT THREAD-SAFE!
+ *
+ * Created by jonn on 8/27/18.
+ */
+public class NioFileCopier {
```

Comment:
Note that if you are doing this to copy from cloud-to-cloud it will be a _lot_ less efficient than using the google storage APIs, as using those APIs doesn't make the data travel through the network to the machine doing the copy. Hopefully this is only used when copying up/down to the cloud.

Reply:
The benefit here is that you can see the progress of your file transfer.

The built-in API methods do not have progress callbacks, so when you want to transfer a 15Gb file, it takes a long time and the user gets very upset because it appears the program is doing nothing.

Reply:
When copying cloud to cloud, making every byte being copied go through your local network computer not only makes the operation take at least 2x longer, but it can also incur egress charges if you're not running the operation on a cloud VM.

This seems like a poor tradeoff to just get a progress display. If you're already streaming from the local computer to the cloud or vice versa then those issues are mitigated and the remaining cost is easier to justify.

Reply:
@pshapiro4broad That's true.  What do you propose to mitigate this?  My specific use case is for a tool that downloads files on the order of 15Gb from our bucket to a user's local running area (wherever that may be).

Are you suggesting that this class should detect when cloud-to-cloud transfers will occur based on the path URL and throw an exception?

Reply:
My point was more about the current use case than about future possible uses, I wanted to make sure that the use case that this is being written for isn't cloud-to-cloud.

One fix would be to check the `Path`s `FileSystemProvider`s. If they're the same, then use `Files.copy(Path, Path)`, otherwise use your code. E.g. call `Path.getFileSystem().provider()`. Providers are singletons based on scheme so you can use `==` to compare them.

Reply:
Since the whole point of the class was to display progress, I'm not going to do additional checking. 

Copying a file using `NioFileCopierWithProgressMeter` should always have the option of displaying a progress meter.